lib/chatclient.js

```diff
@@ -1,20 +1,89 @@
 import { EventEmitter } from 'events';
 
+import pkg from 'websocket';
+const { w3cwebsocket: WebSocket } = pkg;
+
+const WEBSOCKET_URL = 'ws://127.0.0.1:3030';
+
 export default class ChatClient extends EventEmitter {
   constructor(displayName, conversationId) {
     super();
-    // TODO: Implement
+    this.requestCount = 1;
+    this.webSocket = new WebSocket(WEBSOCKET_URL);
+
+    this.webSocket.onopen = () => {
+      console.log('WebSocket Client Connected');
+
+      this.sendEvent('join-conversation', {
+        displayName,
+        conversationId,
+      });
+    };
+
+    this.webSocket.onmessage = (e) => {
+      const data = JSON.parse(e.data);
+
+      if (data.type === 'response' && data.reqId === 1 && data.statusCode === 200) {
+        this.emit('connected');
+      }
+
+      if (data.type === 'event' && data.name === 'new-message') {
+        this.emit('send-message', data);
+      }
+
+      if (data.type === 'response' && data.body && data.statusCode === 200) {
+        this.emit('get-messages', data);
+      }
+    };
+  }
+
+  checkEvent(requestId) {
+    return new Promise((resolve, reject) => {
+      if (requestId === 0) {
+        resolve(true);
+      }
+    });
   }
 
   disconnect() {
-    // TODO: Implement
+    this.emit('disconnected');
+    this.webSocket.close();
+  }
+
+  async sendEvent(commandType, args) {
+    this.webSocket.send(
+      JSON.stringify({
+        command: commandType,
+        reqId: this.requestCount,
+        args,
+      })
+    );
+    this.requestCount++;
   }
 
   getMessages() {
-    // TODO: Implement
+    return new Promise((resolve, reject) => {
```

Comment:
You are going in the right direction here, by mapping request with response. The only feedback here is that you are adding a listener for the same event every time you call `getMessages()`, which results in a memory leak. Plus, since the server does not send responses in the same order as requests, you need to filter the response based on request id, along with message type.

Reply:
Thanks for replying!

I didn't want to spend much time either as this was a time-constrained exercise. So for this PR, I just wanted to lay the foundations and make the tests pass.

You are right, from here there are a few optimizations that could be done, like filtering the responses based on request ID. Another option could be to implement a queue to maintain the order of the messages adding and removing elements from that queue.

I'm also not considering error handling or performance, which is something that could be improved, especially if there are multiple clients and connections.

Thank you again, I really appreciate this feedback!